activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -93,18 +93,34 @@ def clear_query_cache
         end
       end
 
-      def select_all(arel, name = nil, binds = [], preparable: nil)
+      def select_all(arel, name = nil, binds = [], preparable: nil, async: false)
         if @query_cache_enabled && !locked?(arel)
           arel = arel_from_relation(arel)
           sql, binds, preparable = to_sql_and_binds(arel, binds, preparable)
 
-          cache_sql(sql, name, binds) { super(sql, name, binds, preparable: preparable) }
+          if async
+            lookup_sql_cache(sql, name, binds) || super(sql, name, binds, preparable: preparable, async: async)
```

Comment:
I took a conservative approach here. If the cache exist we simply return it, otherwise we don't populate it.

But in theory if we made `FutureResult` respond to all all `Result` methods transparently, we could totally populate the query cache with `FutureResult` instances.